web/auth/interceptors/access_control.go

```diff
@@ -0,0 +1,134 @@
+	"UpdateGroup":             {"teacher"},
+	"DeleteGroup":             {"teacher"},
+	"GetSubmissions":          {"teacher"},
+	"IsEmptyRepo":             {"teacher"},
+	"GetGroupsByCourse":       {"teacher"},
+	"UpdateCourse":            {"teacher"},
+	"UpdateEnrollments":       {"teacher"},
+	"UpdateSubmission":        {"teacher"},
+	"RebuildSubmissions":      {"teacher"},
+	"CreateBenchmark":         {"teacher"},
+	"UpdateBenchmark":         {"teacher"},
+	"DeleteBenchmark":         {"teacher"},
+	"CreateCriterion":         {"teacher"},
+	"UpdateCriterion":         {"teacher"},
+	"DeleteCriterion":         {"teacher"},
+	"CreateReview":            {"teacher"},
+	"UpdateSubmissions":       {"teacher"},
+	"GetReviewers":            {"teacher"},
+	"UpdateAssignments":       {"teacher"},
+	"GetUserByCourse":         {"teacher", "admin"},
+	"GetSubmissionsByCourse":  {"teacher", "admin"},
+	"GetOrganization":         {"admin"},
+	// TODO(vera): GetOrganization and other admin methods not needed in case of admin service
+	// otherwise add all admin methods here
+}
+
+func AccessControl(logger *zap.Logger, tokens *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		// get JWT from context
+		meta, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			logger.Sugar().Errorf("Access control failed: missing metadata")
+			return nil, status.Errorf(codes.Unauthenticated, "access denied")
+		}
+		logger.Sugar().Debugf("Request %s has metadata: %+v", info.FullMethod, meta) // tmp
+		// get claims from jwt
+		for _, c := range meta.Get("cookie") {
+			fields := strings.Fields(c)
+			for _, field := range fields {
+				logger.Sugar().Debugf("metadata field: %s", field) // tmp
+				if strings.Contains(field, "auth") {
+					token := strings.Split(field, "=")[1]
+					logger.Sugar().Debugf("extracted token: %s", token) // tmp
+					claims, err := tokens.GetClaims(token)
+					if err != nil {
+						logger.Sugar().Errorf("Failed to extract claims from JWT: %v", err)
+						return nil, status.Errorf(codes.Unauthenticated, "access denied")
+					}
+					// TODO(vera): refactor this part
+					method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+					// If method is not in the map, there is no restrictions to call it
+					roles, ok := accessRoles[method]
+					logger.Sugar().Debugf("Access control: User %d calls method %s. Expected roles: %v", claims.UserID, method, roles)
+					if ok {
+						for _, role := range roles {
+							switch role {
+							// TODO(vera): refactor case handlers
+							case "user":
+								// Methods that can be accessed by the owner of the UserID can also be accessed by admin.
+								// Skip other checks if user has admin role.
+								// TODO(vera): should better let it continue and check the next role
+								// instead of checking for admin here
+								if !claims.Admin {
```

Comment:
This check and the subsequent switch on the method looks awkward (special cases). I'm not quite sure I understand this check.